server-spi-private/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantType.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.protocol.oidc.grants;
+
+import jakarta.ws.rs.core.HttpHeaders;
+import jakarta.ws.rs.core.MultivaluedHashMap;
+import jakarta.ws.rs.core.MultivaluedMap;
+import jakarta.ws.rs.core.Response;
+
+import java.util.Map;
+
+import org.keycloak.common.ClientConnection;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.http.HttpRequest;
+import org.keycloak.http.HttpResponse;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.provider.Provider;
+import org.keycloak.provider.ProviderFactory;
+import org.keycloak.representations.dpop.DPoP;
+
+/**
+ * Provider interface for OAuth 2.0 grant types
+ *
+ * @author <a href="mailto:demetrio@carretti.pro">Dmitry Telegin</a>
+ */
+public interface OAuth2GrantType extends Provider, ProviderFactory<OAuth2GrantType> {
+
+    /**
+     * Returns the name of the OAuth 2.0 grant type implemented by this provider.
+     * This value will be matched against the "grant_type" token request parameter.
+     *
+     * @return grant type name
+     */
+    String getGrantType();
+
+    /**
+     * Checks if the grant implementation supports the request.
+     * The check will be performed after the initial matching against the "grant_type" parameter.
+     * @param context grant request context
+     * @return request supported
+     */
+    default boolean supports(Context context) {
+        return true;
+    }
+
+    /**
+     * Sets grant request context.
+     * @param context grant request context
+     */
+    void setContext(Context context);
+
+    /**
+     * Processes grant request.
+     *
+     * @return token response
+     */
+    Response process();
+
+    public static class Context {
```

Comment:
+1, I was wondering about the same.

Also I think that most of the fields on the `Context` class can be `final` as they are sent in the constructor (in case we need the fields)

Finally, it can be also good if those fields are `private` and available via getters.

However I also understand that you likely wanted to avoid lots of changes in existing code when accessing those variables and that is why you introduced this?

But at least `final` on the fields can be low-hanging fruit without requiring any major changes IMO?